<?xml version="1.0" ?>
<!-- THIS FILE LAUNCHES THE FRANKA EMIKA PANDA + SOFTHAND: BOTH THE PACKAGES ARE NEEDED -->
<launch>
    <!-- LAUNCH ARGUMENTS -->

    <!-- Panda params -->
    <arg name="robot_ip" default="172.16.0.101"/>
    <arg name="load_gripper" default="false"/>

    <!-- Robot description -->
    <param name="robot_description" command="$(find xacro)/xacro --inorder '$(find adaptive_grasp_controller)/robot/panda_softhand.urdf.xacro'" unless="$(arg load_gripper)" />

    <!-- Panda namespace -->
    <group ns="panda_arm">

        <!-- A copy of robot description -->
        <param name="robot_description" command="$(find xacro)/xacro --inorder '$(find adaptive_grasp_controller)/robot/panda_softhand.urdf.xacro'" unless="$(arg load_gripper)" />

        <!-- Panda hardware interface -->
        <node name="franka_control" pkg="franka_control" type="franka_control_node" output="screen" required="true">
            <rosparam command="load" file="$(find franka_control)/config/franka_control_node.yaml" />
            <param name="robot_ip" value="$(arg robot_ip)" />
        </node>

        <!-- Panda controllers -->
        <rosparam command="load" file="$(find franka_control)/config/default_controllers.yaml" />
        <node name="state_controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="franka_state_controller"/>
        <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen"  args="position_joint_trajectory_controller"/>
        
        <!-- Panda state controllers -->
        <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen"/>
        <node name="joint_state_publisher" type="joint_state_publisher" pkg="joint_state_publisher" output="screen">
            <rosparam if="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states, franka_gripper/joint_states] </rosparam>
            <rosparam unless="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states] </rosparam>
            <param name="rate" value="30"/>
        </node>

        <!-- Panda desired state publisher -->
        <node name="joint_state_desired_publisher" type="joint_state_publisher" pkg="joint_state_publisher" output="screen">
            <rosparam if="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states_desired, franka_gripper/joint_states] </rosparam>
            <rosparam unless="$(arg load_gripper)" param="source_list">[franka_state_controller/joint_states_desired] </rosparam>
            <param name="rate" value="30"/>
            <remap from="/joint_states" to="/joint_states_desired" />
        </node>

    </group>

    <!-- RViz for visualization -->
    <node pkg="rviz" type="rviz" output="screen" name="rviz" args="-d $(find franka_example_controllers)/launch/robot.rviz"/>

</launch>
